README
NOTE FROM CHARLIE MARTÍNEZ (QUIRINUX SO):
This fork is a copy of Modified version based on original version. getweb.in modified
https://foo2zjs.linkevich.net/
   
   ---------------------------------------------------------------------
    Please read the "INSTALL" file to see how to install this software.
    Please read the "INSTALL" file to see how to install this software.
    Please read the "INSTALL" file to see how to install this software.

    Mac OS X users: also read the "INSTALL-osx.txt" file.
    ---------------------------------------------------------------------

GENERAL
    FOO2ZJS
    -------
    foo2zjs converts pbm (B/W) images and 1-bit-per-pixel cmyk images
    (both produced by ghostscript) to Zenographics ZJ-stream format. There
    is some information about the ZJS format at http://ddk.zeno.com.

    With foo2zjs, you can print to some HP and Minolta/QMS ZjStream
    printers, such as these:

	- Konica Minolta magicolor 2430 DL	B/W and color
	- Minolta/QMS 2300 DL			B/W and color
	- Minolta/QMS 2200 DL			B/W and color
	- HP LaserJet P2035			B/W
	- HP LaserJet M1319 MFP			B/W
	- HP LaserJet Pro P1606dn		B/W
	- HP LaserJet Pro P1566			B/W
	- HP LaserJet Pro P1102			B/W
	- HP LaserJet Pro CP1025nw		B/W and color
	- HP LaserJet 1022			B/W
	- HP LaserJet 1020			B/W
	- HP LaserJet 1018			B/W
	- HP LaserJet 1005			B/W
	- HP LaserJet 1000			B/W

    FOO2HP
    ------
    foo2hp converts pbm (B/W) images and N-bit-per-pixel cmyk images
    (both produced by ghostscript) to Zenographics ZJ-stream format. There
    is some information about the ZJS format at http://ddk.zeno.com.

    With foo2hp, you can print to some HP ZjStream printers, such as these:

	- HP Color LaserJet CP1215		B/W and color
	- HP Color LaserJet 1600		B/W and color
	- HP Color LaserJet 2600n		B/W and color

    FOO2LAVA
    --------
    foo2lava converts Ghostscript pbmraw, bitcmyk, or pksmraw  output  for-
    mats  to  monochrome  or color LAVAFLOW or OPL streams, for driving the
    Konica Minolta magicolor 2530 DL network color laser printer, the  Kon-
    ica Minolta magicolor 2480/2480 MF AIO printer, and other Zenographics-
    based LAVAFLOW printers.

	- Konica Minolta magicolor 2530 DL	B/W and color
	- Konica Minolta magicolor 2490 MF	B/W and color
	- Xerox Phaser 6115MFP			B/W and color
	- Konica Minolta magicolor 2480 MF	B/W and color
	- Konica Minolta magicolor 1690MF	B/W and color
	- Konica Minolta magicolor 1680MF	B/W and color
	- Konica Minolta magicolor 1600W	B/W and color

    FOO2OAK
    -------
    foo2oak converts pbm (B/W) images and 1-bit-per-pixel cmyk images
    (both produced by ghostscript) to Oak Technology (now Zoran) OAKT
    format.

    With foo2oak, you can print to the HP LaserJet 1500 and possibly other
    OAKT protocol printers, such as these:

	- HP LaserJet 1500	B/W and color, but only 1-bit per pixel
	- Kyocera KM-2035	B/W
	- Kyocera KM-1635	B/W

    FOO2QPDL
    --------
    foo2qpdl converts Ghostscript pbmraw, bitcmyk, or pksmraw output formats
    to monochrome or color QPDL streams, for driving the Samsung CLP-300,
    CLX-2160, CLP-600, CLX-3160, CLP-610 and the Xerox Phaser 6110
    QPDL printers.

	- Samsung CLP-300	B/W and color
	- Samsung CLP-315	B/W and color
	- Samsung CLP-325	B/W and color
	- Samsung CLP-365	B/W and color
	- Samsung CLP-600	B/W and color
	- Samsung CLP-610	B/W and color
	- Samsung CLP-2160	B/W and color
	- Samsung CLP-3160	B/W and color
	- Samsung CLP-3175	B/W and color
	- Xerox Phaser 6110	B/W and color
	- Xerox Phaser 6110MFP	B/W and color

    FOO2SLX
    -------
    foo2slx converts Ghostscript pbmraw, bitcmyk, or pksmraw output formats
    to monochrome or color SLX streams, for driving the Lexmark C500 net-
    work color laser printer and other SLZ-based printers.  The SLX  stream
    is a variant of ZjStream produced by Software Imaging K.K

	- Lexmark C500n		B/W and color

    FOO2XQX
    -------
    foo2xqx  converts  Ghostscript  pbmraw  to  monochrome XQX streams, for
    driving the HP LaserJet M1005 MFP and other XQX-based printers

	- HP LaserJet M1005 MFP
	- HP LaserJet M1120 MFP
	- HP LaserJet P1005
	- HP LaserJet P1006
	- HP LaserJet P1007
	- HP LaserJet P1008
	- HP LaserJet P1505/P1505n
	- HP LaserJet P2014/P2014n
	- HP LaserJet Pro M1132s MFP
	- HP LaserJet Pro M1212nf MFP

    FOO2HIPERC
    ----------
    foo2hiperc converts Ghostscript pbmraw, bitcmyk, or pksmraw output for-
    mats  to monochrome or color HIPERC streams, for driving the Oki C3200,
    C3300n, C3400n, C5100n, and the C5500n HIPERC printers.

	- Oki C301dn
	- Oki C310dn
	- Oki C511dn
	- Oki C810dn
	- Oki C3100n/C3200n
	- Oki C3300n/C3400n
	- Oki C3530n MFP
	- Oki C5100n/C5150n/C5200n
	- Oki C5500n/C5600n/C5800n

    FOO2HBPL2
    ---------
    foo2hbpl2 converts Ghostscript pbmraw, bitcmyk, or pksmraw output  for‐
    mats  to  monochrome  or  color HBPL version 2 streams.

	- Dell 1355 Multifunction Color Printer
	- Dell C1765 Multifunction Color Printer 
	- Epson AcuLaser CX17NF
	- Epson AcuLaser M1400
	- Fuji Xerox DocuPrint CM205 Multifunction Color Printer
	- Fuji Xerox DocuPrint CM215 Multifunction Color Printer
	- Fuji Xerox DocuPrint M215 
	- Fuji Xerox DocuPrint P205 
	- Xerox Phaser 3010
	- Xerox Phaser 3040
	- Xerox WorkCentre 3045 Multifunction Monochrome Printer 
	- Xerox WorkCentre 6015 Multifunction Color Printer 

Please read the manual pages for more details on the program operation.

    $ acroread manual.pdf

BUGS AND DEFICIENCIES
    - Need to tune the manufacturer-supplied ICM color profiles to match the
      characteristics of Ghostscript.

    - Need to figure out how to get near-invisible dithering out of Ghostscript
      without affecting the colors.  With GS 7.05, if you set -dDITHERING=150,
      for example, the dithering becomes less visible but you get areas of
      color blotches.  No way to win.

    - Ghostscript 8.0 has worse color dithering compared to 7.05.  The symptom
      is slanted bands separated by 0.5 to 2mm, at a 30 degree angle.

    - Need to support a better input color file format which includes
      explicit page boundary indications.  The current use of the Ghostscript
      "bitcmyk" format precludes having pages of different sizes within a
      Postscript stream.

    - It is not entirely clear whether the JBIG-KIT compression library
      produces compressed data compatible with all implementations of JBIG
      in ZjStream printers.  It is possible that bugs or subtle
      implementation differences might result in incorrect printer behavior.

AUTHORS
    The foo2zjs engine program began life as Robert Szalai's 'pbmtozjs'
    program.  It also uses Markus Kuhn's JBIG-KIT compression library
    (included, but available at http://www.cl.cam.ac.uk/~mgk25/jbigkit/),
    and Marti Maria's icc2ps program (included, but also available at
    http://www.littlecms.com).

    The foo2zjs engine was overhauled by Rick Richardson to add color
    support for the Minolta/QMS 2300DL, add command line options, limit
    data chunk size to 65536 bytes, and other miscellaneous features.
    Rick also contributed the zjsdecode program and the foomatic database
    entries for the Minolta/QMS magicolor 2300 DL.

    The foo2* engine was written entirely from scratch by Rick Richardson.

    The latest version of foo2zjs can be found at:

	http://foo2zjs.rkkda.com
	wget http://foo2zjs.rkkda.com/foo2zjs.tar.gz

You can contact the current author at mailto://rick.richardson@comcast.net
I do not read or reply to HTML email or messages sent under an alias.

LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or (at
your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

If you want to use this program under different license conditions,
then contact the author for an arrangement.

PATENTS
It is possible that certain products which can be built using the jbig
software module might form inventions protected by patent rights in
some countries (e.g., by patents about arithmetic coding algorithms
owned by IBM and AT&T in the USA). Provision of this software by the
author does NOT include any licences for any patents. In those
countries where a patent licence is required for certain applications
of this software module, you will have to obtain such a licence
yourself.

"So, as far as I understand currently (IANAPL), JBIG1 will be free of
known patents in the United States from 2012-04-04 onwards, at the
very latest, and outside the United States, JBIG1 will be patent free
already from 2011-02-25." - Markus Kuhn


ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default
	* hbpldecode: Print out paper codes.

2014-03-15	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More changes from Dave.  Still gets 
	    Segmentation fault (core dumped)
	* hbpldecode: Changed unsigned to int in 2 places.  Works good!!!

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Bug with no -d/-r.

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Changes to pick apart version 1 from
	    Dave Coffin <dcoffin@cybercom.net>
	* I get:
	    Decoding page 1 to page-01.ppm ...
	    Segmentation fault (core dumped)

2014-03-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint M215 MFP using the foo2hbpl2 driver.

2014-02-09	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox WorkCentre 3045 MFP using the foo2hbpl2 driver.

2014-01-26	Rick Richardson <rick.richardson@comcast.net>
	* foo2qpdl et al: with -z3 (e.g. CLP-365) you can adjust
	    Brightness and Contrast for monochrome prints, and you can adjust
	    Saturation, Cyan-Red, Magenta-Green, Yellow-Blue Balances.
	* foo2hp: somewhere a bug in foomatic (probably 4.x) crept in
	    so foo2hp-AlignCMYK.xml was only processing one option instead
	    of three options!  Split the file into foo2hp-AlignC.xml,
	    foo2hp-AlignM.xml, and foo2hp-AlignY.xml.

2014-01-24	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Dell C1765nf/nfw using the foo2hbpl2 driver.

2014-01-08	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser M1400 using the foo2hbpl2 driver.

2013-12-25	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: HP LaserJet Pro M1212nf MFP using the foo2xqx driver.

2013-12-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc: clean up man pages

2013-11-29	Rick Richardson <rick.richardson@comcast.net>
	* foo2hiperc-wrapper, foomatic and PPD stuff:
	    add -z1 for Oki C810 so everything will be

xqx.h
/*
 * Manifest constants for the ZjStream protocol
 *
 * I'm told that a lot of this stuff came from a file called "zjrca.h".
 * But a Google search does not turn up that file. I added my own
 * improvements and missing pieces. -Rick
 */

#include <inttypes.h>
typedef uint32_t	DWORD;
typedef uint16_t	WORD;
typedef uint8_t		BYTE;

typedef enum {
    XQX_START_DOC	= 1,
    XQX_END_DOC		= 2,
    XQX_START_PAGE	= 3,
    XQX_END_PAGE	= 4,
    XQX_START_PLANE	= 5,
    XQX_END_PLANE	= 6,
    XQX_JBIG		= 7,

    // XQX_2600N_PAUSE	=11,
    // XQX_2600N	=12,
} XQX_TYPE;

typedef struct _XQX_HEADER {
    DWORD type;           /*  ZJ_TYPE */
    DWORD items;          /*  total record size, includes sizeof(ZJ_HEADER) */
} XQX_HEADER;

typedef enum {
    XQXI_DMDUPLEX	= 0x10000002,	// from DEVMODE (DMBIN?)
    XQXI_COPIES		= 0x20000005,	// from DEVMODE (DMBIN?)
    XQXI_DMDEFAULTSOURCE= 0x20000006,	// from DEVMODE (DMBIN?)
    XQXI_DMMEDIATYPE	= 0x20000000,	// from DEVMODE
    XQXI_RESOLUTION_X	= 0x20000008,	// dots per inch
    XQXI_RESOLUTION_Y	= 0x20000009,
    XQXI_RASTER_X	= 0x2000000d,	// dots per inch
    XQXI_RASTER_Y	= 0x2000000e,
    XQXI_VIDEO_BPP	= 0x2000000a,	// video bits per pixel
    XQXI_VIDEO_X	= 0x2000000f,	// video dimensions
    XQXI_VIDEO_Y	= 0x20000010,
    XQXI_ECONOMODE	= 0x20000011,	// HP's "EconoMode", 0=OFF, 1=ON
    XQXI_DMPAPER	= 0x20000001,
    XQXI_DUPLEX_PAUSE	= 0x20000012,	// 0,1=front side; 2-3=back side
    XQXI_BIH		= 0x40000002,
    XQXI_END		= 0x80000001,

    ZJI_LAST
} XQX_ITEM;

typedef enum
{
    DMDUPLEX_OFF	= 1,
    DMDUPLEX_LONGEDGE	= 2,
    DMDUPLEX_SHORTEDGE	= 3,
    DMDUPLEX_MANUALLONG	= 4,
    DMDUPLEX_MANUALSHORT= 5
} DMDUPLEX;

typedef enum {
    DMBIN_TRAY1		= 1,
    DMBIN_MANUAL	= 4,
    DMBIN_AUTO		= 7,
} DM_BIN;

typedef enum {
    DMMEDIA_UNSPECIFIED		= 0,	// Unspecified
    DMMEDIA_PLAIN		= 1,	// Plain paper
    DMMEDIA_HP_MULTIPURPOSE	= 301,
    DMMEDIA_HP_OFFICE		= 300,
    DMMEDIA_HP_ALL_IN_ONE	= 303,
    DMMEDIA_HP_LASERJET		= 302,
    DMMEDIA_HP_PREMIUM_LASERJET	= 304,
    DMMEDIA_PREPRINTED		= 514,
    DMMEDIA_LETTERHEAD		= 513,	// Letterhead
    DMMEDIA_TRANSPARENCY	= 2,	// Transparency
    DMMEDIA_PREPUNCHED		= 515,
    DMMEDIA_LABELS		= 265,	// Labels
    DMMEDIA_BOND		= 260,
    DMMEDIA_RECYCLED		= 516,
    DMMEDIA_COLOR		= 512,
    DMMEDIA_LIGHT		= 258,
    DMMEDIA_HEAVY		= 262,
    DMMEDIA_CARDSTOCK		= 261,
    DMMEDIA_ENVELOPE		= 267,	// Envelope
    DMMEDIA_ROUGH		= 263,
} DMMEDIA;

typedef enum {
    DMCOLOR_MONOCHROME	= 1,
    DMCOLOR_COLOR	= 2,
} DMCOLOR;

typedef enum {
    DMORIENT_PORTRAIT	= 1,
    DMORIENT_LANDSCAPE	= 2,
} DMORIENT;

typedef enum {
    DMPAPER_LETTER	= 1,	// Letter, 8 1/2- by 11-inches
    DMPAPER_LETTERSMALL	= 2,	// Letter Small, 8 1/2- by 11-inches
    DMPAPER_TABLOID	= 3,	// Tabloid, 11- by 17-inches
    DMPAPER_LEDGER	= 4,	// Ledger, 17- by 11-inches
    DMPAPER_LEGAL	= 5,	// Legal, 8 1/2- by 14-inches
    DMPAPER_STATEMENT	= 6,	// Statement, 5 1/2- by 8 1/2-inches
    DMPAPER_EXECUTIVE	= 7,	// Executive, 7 1/4- by 10 1/2-inches
    DMPAPER_A3		= 8,	// A3 sheet, 297- by 420-millimeters
    DMPAPER_A4		= 9,	// A4 Sheet, 210- by 297-millimeters
    DMPAPER_A4SMALL	=10,	// A4 small sheet, 210- by 297-millimeters
    DMPAPER_A5		=11,	// A5 sheet, 148- by 210-millimeters
    DMPAPER_B4		=12,	// B4 sheet, 250- by 354-millimeters
    DMPAPER_B5		=13,	// B5 sheet, 182- by 257-millimeter paper
    DMPAPER_FOLIO	=14,	// Folio, 8 1/2- by 13-inch paper
    DMPAPER_QUARTO	=15,	// Quarto, 215- by 275-millimeter paper
    DMPAPER_10X14	=16,	// 10- by 14-inch sheet
    DMPAPER_11X17	=17,	// 11- by 17-inch sheet
    DMPAPER_NOTE	=18,	// Note, 8 1/2- by 11-inches
    DMPAPER_ENV_9	=19,	// #9 Envelope, 3 7/8- by 8 7/8-inches
    DMPAPER_ENV_10	=20,	// #10 Envelope, 4 1/8- by 9 1/2-inches
    DMPAPER_ENV_11	=21,	// #11 Envelope, 4 1/2- by 10 3/8-inches
    DMPAPER_ENV_12	=22,	// #12 Envelope, 4 3/4- by 11-inches
    DMPAPER_ENV_14	=23,	// #14 Envelope, 5- by 11 1/2-inches
    DMPAPER_CSHEET	=24,	// C Sheet, 17- by 22-inches
    DMPAPER_DSHEET	=25,	// D Sheet, 22- by 34-inches
    DMPAPER_ESHEET	=26,	// E Sheet, 34- by 44-inches
    DMPAPER_ENV_DL	=27,	// DL Envelope, 110- by 220-millimeters
    DMPAPER_ENV_C5	=28,	// C5 Envelope, 162- by 229-millimeters
    DMPAPER_ENV_C3	=29,	// C3 Envelope, 324- by 458-millimeters
    DMPAPER_ENV_C4	=30,	// C4 Envelope, 229- by 324-millimeters
    DMPAPER_ENV_C6	=31,	// C6 Envelope, 114- by 162-millimeters
    DMPAPER_ENV_C65	=32,	// C65 Envelope, 114- by 229-millimeters
    DMPAPER_ENV_B4	=33,	// B4 Envelope, 250- by 353-millimeters
    DMPAPER_ENV_B5	=34,	// B5 Envelope, 176- by 250-millimeters
    DMPAPER_ENV_B6	=35,	// B6 Envelope, 176- by 125-millimeters
    DMPAPER_ENV_ITALY	=36,	// Italy Envelope, 110- by 230-millimeters
    DMPAPER_ENV_MONARCH	=37,	// Monarch Envelope, 3 7/8- by 7 1/2-inches
    DMPAPER_ENV_PERSONAL=38,	// 6 3/4 Envelope, 3 5/8- by 6 1/2-inches
    DMPAPER_FANFOLD_US	=39,	// US Std Fanfold, 14 7/8- by 11-inches
    DMPAPER_FANFOLD_STD_GERMAN	=40,	// German Std Fanfold, 8 1/2 x 12 in
    DMPAPER_FANFOLD_LGL_GERMAN	=41,	// German Legal Fanfold, 8 1/2 x 13 in

    DMPAPER_ISO_B4		   =42,	// B4 (ISO) 250 x 353 mm
    DMPAPER_JAPANESE_POSTCARD	   =43,	// Japanese Postcard 100 x 148 mm
    DMPAPER_9X11		   =44,	// 9 x 11 in
    DMPAPER_10X11		   =45,	// 10 x 11 in
    DMPAPER_15X11		   =46,	// 15 x 11 in
    DMPAPER_ENV_INVITE		   =47,	// Envelope Invite 220 x 220 mm
    DMPAPER_RESERVED_48		   =48,	// RESERVED--DO NOT USE
    DMPAPER_RESERVED_49		   =49,	// RESERVED--DO NOT USE
    DMPAPER_LETTER_EXTRA	   =50,	// Letter Extra 9 \275 x 12 in
    DMPAPER_LEGAL_EXTRA		   =51,	// Legal Extra 9 \275 x 15 in
    DMPAPER_TABLOID_EXTRA	   =52,	// Tabloid Extra 11.69 x 18 in
    DMPAPER_A4_EXTRA		   =53,	// A4 Extra 9.27 x 12.69 in
    DMPAPER_LETTER_TRANSVERSE	   =54,	// Letter Transverse 8 \275 x 11 in
    DMPAPER_A4_TRANSVERSE	   =55,	// A4 Transverse 210 x 297 mm
    DMPAPER_LETTER_EXTRA_TRANSVERSE=56,	// Letter Extra Transverse 9\275 x 12 in
    DMPAPER_A_PLUS		   =57,	// SuperA/SuperA/A4 227 x 356 mm
    DMPAPER_B_PLUS		   =58,	// SuperB/SuperB/A3 305 x 487 mm
    DMPAPER_LETTER_PLUS		   =59,	// Letter Plus 8.5 x 12.69 in
    DMPAPER_A4_PLUS		   =60,	// A4 Plus 210 x 330 mm
    DMPAPER_A5_TRANSVERSE	   =61,	// A5 Transverse 148 x 210 mm
    DMPAPER_B5_TRANSVERSE	   =62,	// B5 (JIS) Transverse 182 x 257 mm
    DMPAPER_A3_EXTRA		   =63,	// A3 Extra 322 x 445 mm
    DMPAPER_A5_EXTRA		   =64,	// A5 Extra 174 x 235 mm
    DMPAPER_B5_EXTRA		   =65,	// B5 (ISO) Extra 201 x 276 mm
    DMPAPER_A2			   =66,	// A2 420 x 594 mm
    DMPAPER_A3_TRANSVERSE	   =67,	// A3 Transverse 297 x 420 mm
    DMPAPER_A3_EXTRA_TRANSVERSE	   =68,	// A3 Extra Transverse 322 x 445 mm
} DMPAPER;

typedef struct _XQX_ITEM_HEADER {
    DWORD type;		// type
    DWORD size;		// total record size following
} XQX_ITEM_HEADER;

typedef struct _XQX_ITEM_UINT32 {
    XQX_ITEM_HEADER	header;
    DWORD		value;
} XQX_ITEM_UINT32;

typedef union _SWAP_32{
    char		byte[sizeof(uint32_t)];
    uint32_t		dword;
} SWAP_32;

typedef union _SWAP_16{
    char		byte[sizeof(uint16_t)];
    uint16_t		word;
} SWAP_16;

static inline uint32_t
be32(uint32_t dword)
{
    SWAP_32	swap;
    uint32_t	probe = 1;

    if (((char *)&probe)[0] == 1)
    {
	swap.byte[3] = (( SWAP_32 )dword).byte[0]; 
	swap.byte[2] = (( SWAP_32 )dword).byte[1];
	swap.byte[1] = (( SWAP_32 )dword).byte[2]; 
	swap.byte[0] = (( SWAP_32 )dword).byte[3];
	return swap.dword;
    }
    else
	return dword;
}

static inline uint16_t
be16(uint16_t word)
{
    SWAP_16	swap;
    uint16_t	probe = 1;

    if (((char *)&probe)[0] == 1)
    {
	swap.byte[1] = (( SWAP_16 )word).byte[0]; 
	swap.byte[0] = (( SWAP_16 )word).byte[1];
	return swap.word;
    }
    else
	return word;
}
